libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -1483,454 +2019,575 @@ IotMqttError_t _IotMqtt_DeserializePuback( _mqttPacket_t * pPuback )
+                                            size_t bufferSize )
 {
-    /* PINGREQ packets are always the same. */
-    static const uint8_t pPingreq[ MQTT_PACKET_PINGREQ_SIZE ] =
+    IOT_FUNCTION_ENTRY( IotMqttError_t, IOT_MQTT_SUCCESS );
+    uint8_t * pDisconnectPacket = NULL;
+    size_t remainingLength = 0;
+
+    if( pBuffer == NULL )
     {
-        MQTT_PACKET_TYPE_PINGREQ,
-        0x00
-    };
+        IotLogError( "IotMqtt_SerializeDisconnect() called with NULL buffer pointer." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
+    }
 
-    /* Set the output parameters. */
-    *pPingreqPacket = ( uint8_t * ) pPingreq;
-    *pPacketSize = MQTT_PACKET_PINGREQ_SIZE;
+    if( bufferSize < MQTT_PACKET_DISCONNECT_SIZE )
+    {
+        IotLogError( "Disconnect packet length (%lu) exceeds buffer size (%lu).",
+                     MQTT_PACKET_DISCONNECT_SIZE, bufferSize );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
+    }
 
-    /* Print out the PINGREQ packet for debugging purposes. */
-    IotLog_PrintBuffer( "MQTT PINGREQ packet:", pPingreq, MQTT_PACKET_PINGREQ_SIZE );
+    /* Call internal function with local varaibles, as disconnect  uses
+     * static memory, there is no need to pass the buffer
+     * Note: _IotMqtt_SerializeDisconnect always succeeds */
+    _IotMqtt_SerializeDisconnect( &pDisconnectPacket, &remainingLength );
 
-    return IOT_MQTT_SUCCESS;
+    memcpy( pBuffer, pDisconnectPacket, MQTT_PACKET_DISCONNECT_SIZE );
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
 }
 
 /*-----------------------------------------------------------*/
 
-IotMqttError_t _IotMqtt_DeserializePingresp( _mqttPacket_t * pPingresp )
+IotMqttError_t IotMqtt_SerializePingreq( uint8_t * pBuffer,
+                                         size_t bufferSize )
 {
     IOT_FUNCTION_ENTRY( IotMqttError_t, IOT_MQTT_SUCCESS );
+    uint8_t * pPingreqPacket = NULL;
+    size_t packetSize = 0;
 
-    /* Check that the control packet type is 0xd0. */
-    if( pPingresp->type != MQTT_PACKET_TYPE_PINGRESP )
+    if( pBuffer == NULL )
     {
-        IotLog( IOT_LOG_ERROR,
-                &_logHideAll,
-                "Bad control packet type 0x%02x.",
-                pPingresp->type );
-
-        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_RESPONSE );
+        IotLogError( "IotMqtt_SerializePingreq() called with NULL buffer pointer." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
     }
-    else
+
+    if( bufferSize < MQTT_PACKET_PINGREQ_SIZE )
     {
-        EMPTY_ELSE_MARKER;
+        IotLogError( "Pingreq length (%lu) exceeds buffer size (%lu).",
+                     MQTT_PACKET_DISCONNECT_SIZE, bufferSize );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
     }
 
-    /* Check the "Remaining length" (second byte) of the received PINGRESP. */
-    if( pPingresp->remainingLength != MQTT_PACKET_PINGRESP_REMAINING_LENGTH )
+    /* Call internal function with local varaibles, as ping request uses
```

Comment:
spelling: _variables_, also search for the misspelled version as I see multiple occurrences.